File: Sean的leetcode/python/3Sum Closest.py
###
# Time Complexity: O(nlogn+n^2)
# Space Complexity: O(n) for sorting
###
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if len(nums) < 3:
            return -1
        mindis = float('inf')
        result = None
        nums.sort()

        for i in xrange(len(nums) - 2):
            l = i + 1
            r = len(nums) - 1
            while l < r:
                s = nums[i] + nums[l] + nums[r]
                if s == target:
                    return s
                elif s < target:
                    if abs(s - target) < mindis:
                        mindis = abs(s - target)
                        result = s
                    l += 1
                else:
                    if abs(s - target) < mindis:
                        mindis = abs(s - target)
                        result = s
                    r -= 1
        return result

File: MyLeetCode/Leetcode/3Sum Closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int min_diff_sum = INT_MAX;
        int min_diff = INT_MAX;
        int n = nums.size();

        sort(nums.begin(), nums.end());

        for (int i = 0; i < n - 2; i++)
        {
            for (int left = i + 1, right = n - 1; left < right;)
            {
                int sum = nums[i] + nums[left] + nums[right];

                if (abs(sum - target) < min_diff)
                {
                    min_diff = abs(sum - target);
                    min_diff_sum = sum;
                }

                if (sum == target)
                    return sum;
                else if (sum > target)
                    right--;
                else
                    left++;
            }
            while (i + 1 < n - 2 && nums[i+1] == nums[i])
                i++;
        }
        return min_diff_sum;
    }
};